### PR TITLE
Unify wrapper paths

### PR DESCRIPTION
**Pull request by compnerd**
Merge the block and non-blocking sys call wrappers

### Motivation:

The two paths are nearly identical with the core difference being that blocking calls ignore `EWOULDBLOCK`.  This simplifies the handling for the different types by adding a parameter to the `wrapSyscall` method.

### Modifications:

Added a new parameter `nonblocking` to the `wrapSyscall` and `wrapSyscallWouldBlock`.  Those paths are then merged into a single `call` method.

### Result:

No external changes.

### CODE REVIEW

File: Sources/NIO/System.swift [hunk 1]
```diff
@@ -113,43 +113,32 @@ private func preconditionIsNotBlacklistedErrno(err: CInt, where function: String
...
-internal func wrapSyscall<T: FixedWidthInteger>(where function: String = #function, _ body: () throws -> T) throws -> T {
+internal func call<T: FixedWidthInteger>(nonblocking: Bool, where function: String = #function, _ body: () throws -> T) throws -> IOResult<T> {
     while true {
         let res = try body()
         if res == -1 {
             let err = errno
-            if err == EINTR {
-                continue
-            }
+            if err == EINTR { continue }
```
Comment (weissi) [inline Sources/NIO/System.swift]: nit: we usually put newlines after every `{`

File: Sources/NIO/System.swift [hunk 2]
```diff
@@ -113,43 +113,32 @@ private func preconditionIsNotBlacklistedErrno(err: CInt, where function: String
...
+      }
+    }
+  }
 }
 
 /* Sorry, we really try hard to not use underscored attributes. In this case however we seem to break the inlining threshold which makes a system call take twice the time, ie. we need this exception. */
 @inline(__always)
-@discardableResult
-internal func wrapSyscall<T: FixedWidthInteger>(where function: String = #function, _ body: () throws -> T) throws -> T {
+internal func call<T: FixedWidthInteger>(nonblocking: Bool, where function: String = #function, _ body: () throws -> T) throws -> IOResult<T> {
```
Comment (weissi) [inline Sources/NIO/System.swift]: I think we should still name this `wrapSyscall` or `callSyscall` or something. This wrapper only works if the called function behaves like a unix syscall (ie. -1 == error)

**Comment by compnerd** [comment]
I think addressing the generated code issue first is more important, fixing the style stuff can come later.  I checked on an ubuntu build, it depends on how you write the code and what you want the result to be.

Looking at the code for:

```swift
while true {
  let res = try body()
  if res == -1 {
    let err = errno
    if err == EINTR { continue }
    if blocking && err == EWOULDBLOCK { return .wouldBlock(0) }
    preconditionIsNotBlacklistedErrno(err: err, where: function)
    throw IOError(errnoCode: err, reason: function)
  }
  return .processed(res)
}
```

generates the same assembly (modulo offsets) - I've attached the before and after results.

However, it generates two functions, the static function and the merged function.

Writing it as:

```swift
while true {
  let res = try body()
  if res == -1 {
    let err = errno
    switch err {
    case EINTR: continue
    case EWOULDBLOCK:
      if blocking { return .wouldBlock(0) }
      fallthrough
    default:
      preconditionIsNotBlacklistedErrno(err: err, where: function)
      throw IOError(errnoCode: err, reason: function)
    }
  }
  return .processed(res)
}
```
produces a single function instead.

The single function is a little bit heavier on the frame size, but overall reduces the amount of code.  There is a minor amount of extra overhead (2 extra `cmp`+`jne`, 4 total instructions) due to the unnecessary precondition check in the case of `EWOULDBLOCK`, but that is not going to happen anyways - and you can see that with `-Ounchecked` results.

**Comment by compnerd** [comment]
[after.txt](https://github.com/apple/swift-nio/files/4285073/after.txt)
[before.txt](https://github.com/apple/swift-nio/files/4285074/before.txt)
[variant.txt](https://github.com/apple/swift-nio/files/4285075/variant.txt)

**Comment by Lukasa** [comment]
@compnerd Are you still ok with us going ahead and merging this separate from #1440?

**Comment by compnerd** [comment]
Yes, that would be wonderful.  Small steps make it easier :)

**Pull request merged**